Built-in assembler for debugger interface

vyuuui opened the pull request:
A feature I've wanted in Dolphin's debugger for a while was some form of integrated assembler. Right now, the only options when trying to do live patches are either using a third party tool to generate a Gecko code, or extracting single machine code words from the output of an assembler and patching each instruction individually in the Code Viewer.

With this change, I'm hoping to introduce two useful features to speed up the old process which will also justify the creation of an entirely handwritten assembler for Gekko being shoved into Dolphin's codebase.

### Assembler Widget
A new docked widget included in the Debugger UI (alongside codeview, regview, etc...). It acts much like an IDE for assembly. Below is a catalogue of its primary feature set
 - Multi-line assembly
 - Syntax highlighting and parse error detection in-editor.
 - Directly applying raw output to RAM
 - Multiple output formats (Raw, AR, Gecko execute, Gecko trampoline)
 - Saving and loading files

[Here's a sample video](https://youtu.be/7A5cqR4LU6k)
Clean themed tab bar icons were designed and illustrated by [Mr. Mendelli](https://github.com/MrMendelli)

### Inline Assembly Patcher
An extra context menu option in the instruction pane within the CodeView widget. It supports assembling a single line, and should make the process of single instruction patches much less annoying. It will initially be filled with the raw machine code of the instruction being patched. The UI will display any errors with the entered assembly, updating on each character typed. It should be relatively self-explanatory.

https://user-images.githubusercontent.com/92327340/215096250-a3d43e70-c146-49f6-bd00-2d918c80ee80.mp4

### Assembler Language
It would also be pertinent to include some documentation for the language itself. Since it's relatively simple, [this markdown document](https://gist.github.com/vyuuui/05307e1c9fc96d288c1a333a00458fba#file-assembler-documentation-md) should be good enough.

Simonx22 commented:
@dolphin-emu-bot rebuild

vyuuui commented:
Whoops, forgot to run clang-format on this before submitting... I'll split those into three squash-able commits for now.

hzzzr commented:
assembly

Pokechu22 commented:
@dolphin-emu-bot rebuild

Pokechu22 commented:
You've currently got compile errors (missing includes of e.g. `QStringView`), and also formatting errors (possibly caused by a different version of `clang-format`? I'm not entirely sure how to set it up).

Regarding instruction replacement: I think it'd be more useful to pre-populate that with the existing disassembled instruction. The existing instruction patching functionality shows the disassembly of the hex you've typed in, so that probably could be used instead (although I'm not sure if it's always going to be unambiguous enough to produce the same instruction, and there's also a chance that the disassembler produces different results from what the assembler expects).

vyuuui commented:
> Regarding instruction replacement: I think it'd be more useful to pre-populate that with the existing disassembled instruction. The existing instruction patching functionality shows the disassembly of the hex you've typed in, so that probably could be used instead (although I'm not sure if it's always going to be unambiguous enough to produce the same instruction, and there's also a chance that the disassembler produces different results from what the assembler expects).

It's as you said -- I would need to modify parts of the disassembler, as the disassembly of some instructions don't match to what the assembler would expect. Changing the disassembler just for that might be outside the purview of this feature, but I'm not against it.

vyuuui commented:
Build errors should be dealt with, needed to get it working on windows as well.

vyuuui commented:
... As well as fixing the code style. That should be it for now.

Pokechu22 commented:
@dolphin-emu-bot rebuild

Pokechu22 commented:
There still seem to be some build issues on non-Windows platforms. I think they can be fixed by adding `#include <optional>`, `<vector>`, `<tuple>`, `<utility>` to `GekkoSyntaxHighlight.h`, and (I'm less sure of this) by moving the `#include <filesystem>` before `#include <QFile>` in AssemblyEditor.cpp.

vyuuui commented:
Agh, quite annoying. I'll try building from a fresh copy of debian, some transitive include with my installed packages is letting this slip by.

lioncash commented on Source/Core/DolphinQt/Debugger/GekkoSyntaxHighlight.h:
```diff
@@ -0,0 +1,53 @@
+  std::optional<Common::GekkoAssembler::AssemblerError> error;
+  bool error_at_eol = false;
+
+  virtual ~BlockInfo() {}
```
This can be omitted

lioncash commented on Source/Core/DolphinQt/Debugger/GekkoSyntaxHighlight.h:
```diff
@@ -0,0 +1,53 @@
+  Q_OBJECT;
+
+public:
+  GekkoSyntaxHighlight(QTextDocument* document) : QSyntaxHighlighter(document), m_mode(0) {}
```
m_mode can be initialized directly. m_cursor_loc should also have a default initial value.

lioncash reviewed (commented):
Just did a cursory rundown of everything

vyuuui commented on Source/Core/Common/Assembler/AssemblerTables.cpp:
```diff
@@ -0,0 +1,1491 @@
+  overfill = count == MAX_OPERANDS;
+  for (size_t i = before + 1; i <= count && i < MAX_OPERANDS; i++)
+  {
+    auto tmp = list[i];
```
Got that, as well as a few other places that did a pointless XOR swap.

vyuuui commented on Source/Core/Common/Assembler/AssemblerTables.h:
```diff
@@ -0,0 +1,149 @@
+};
+
+// Mapping of SPRG names to values
+extern CaseInsensitiveDict<u32, '_'> sprg_map;
```
Honestly wasn't aware you could have const globals, since const is supposed to have static linkage. Added to all global 'tables' here.

vyuuui commented on Source/Core/Common/Assembler/CaseInsensitiveDict.h:
```diff
@@ -0,0 +1,126 @@
+class CaseInsensitiveDict
+{
+public:
+  CaseInsensitiveDict(std::initializer_list<std::pair<std::string_view, V>> const& il)
```
Hopefully caught every occurrence of it.

vyuuui commented on Source/Core/DolphinQt/Debugger/GekkoSyntaxHighlight.cpp:
```diff
@@ -0,0 +1,243 @@
+    }
+  }
+  else if (m_mode == 2)
+  {
```
Accidentally left it in from earlier, removed.

Simonx22 commented:
@dolphin-emu-bot rebuild

vyuuui commented:
Sorry, need to rebase this onto master as well.

Simonx22 commented:
@dolphin-emu-bot rebuild

vyuuui commented:
Nevermind, anything that was using that set already needed to convert string views to strings anyway, so I'll just roll with set using the default comparator. That should fix the android build failure.

Pokechu22 commented:
@dolphin-emu-bot rebuild

hzzzr commented:
LGTM!

hzzzr commented:
When will this be merged?  I am looking forward to using this interface for my new assembly project 🚀

Pokechu22 commented:
It'll probably be a bit because I want to take some time to review it in more detail. If you want to try it out for now, there's a windows build at https://dl.dolphin-emu.org/prs/39/1f/pr-11497-dolphin-latest-x64.7z.

notscimmy commented:
Approved! LGTM! Merge! Ship it!

Pokechu22 commented:
This seems really neat, but I still haven't had time to review it in detail (and there's a merge conflict now).

As a thought, would it be possible to pre-populate the text field with the disassembled instruction after verifying that the disassembled instruction re-assembles back to the same instruction? I feel like _most_ instructions should be able to round-trip properly (with branch instructions being the most likely exception), and that would make edits a lot easier.

vyuuui commented:
There's no rush for me to get this in, I'll deal with that merge conflict.

> As a thought, would it be possible to pre-populate the text field with the disassembled instruction after verifying that the disassembled instruction re-assembles back to the same instruction?

That could work, it would only really break on branch and SIMD instructions. Could try massaging the disassembly of branch, but for SIMD I think there's too many cases of unexpected output (registers prefixed with `p`, formatting operands based on the operation).

TryTwo commented:
@vyuuui  Are you still around?

vyuuui commented:
Thanks for reminding me, forgot to update this a while ago. Was missing the changes to accessing core, this should build now.

TryTwo commented:
@dolphin-emu-bot rebuild

TryTwo commented:
Thanks!  Would it be okay to put this in #12312? It might get more testing feedback.

vyuuui commented:
Sure, go ahead. This still hasn't been reviewed, so it's likely going to need some changes.

Tilka commented:
@vyuuui A very welcome addition!

Could you add license headers to all the new files, please? Just copy-paste from an existing file and replace the year.

sepalani commented on Source/Core/Common/Assembler/GekkoLexer.h:
```diff
@@ -0,0 +1,185 @@
+
+namespace Common::GekkoAssembler::detail
+{
+void ConvertStringLiteral(std::string_view literal, std::vector<u8>& out_vec);
```
The out parameter should be passed as pointer based on [Dolphin coding style](https://github.com/dolphin-emu/dolphin/blob/master/Contributing.md#cpp-code-functions).

sepalani commented on Source/Core/Common/Assembler/GekkoLexer.h:
```diff
@@ -0,0 +1,185 @@
+  void EatAndReset();
+
+  template <size_t N>
+  void LookaheadTaggedN(std::array<Tagged<CursorPosition, AssemblerToken>, N>& tokens_out) const
+  {
+    const size_t filled_amt = std::min(m_lexed_tokens.size(), N);
+
+    std::copy_n(m_lexed_tokens.begin(), filled_amt, tokens_out.begin());
+
+    std::generate_n(tokens_out.begin() + filled_amt, N - filled_amt, [this] {
+      CursorPosition p = m_pos;
+      return m_lexed_tokens.emplace_back(p, LexSingle());
+    });
+  }
+
+  template <size_t N>
+  void LookaheadN(std::array<AssemblerToken, N>& tokens_out) const
```
Ditto regarding the out parameters which should be passed as pointer.

sepalani commented on Source/Core/Common/Assembler/GekkoLexer.cpp:
```diff
@@ -0,0 +1,791 @@
+#include "Common/Assembler/GekkoLexer.h"
```
Missing license header.

```suggestion
// Copyright 2023 Dolphin Emulator Project
// SPDX-License-Identifier: GPL-2.0-or-later

#include "Common/Assembler/GekkoLexer.h"
```

sepalani commented on Source/Core/Common/Assembler/GekkoIRGen.cpp:
```diff
@@ -0,0 +1,829 @@
+
+  case GekkoDirective::Skip:
+  {
+    u32 skip_len = static_cast<u32>(m_eval_stack.back());
```
```suggestion
    const u32 skip_len = static_cast<u32>(m_eval_stack.back());
```

sepalani commented on Source/Core/Common/Assembler/GekkoIRGen.cpp:
```diff
@@ -0,0 +1,829 @@
+
+void GekkoIRPlugin::OnLabelDecl(std::string_view name)
+{
+  std::string name_str(name);
```
```suggestion
  const std::string name_str(name);
```

sepalani commented on Source/Core/Common/Assembler/GekkoIRGen.cpp:
```diff
@@ -0,0 +1,829 @@
+
+void GekkoIRPlugin::OnVarDecl(std::string_view name)
+{
+  std::string name_str(name);
```
```suggestion
  const std::string name_str(name);
```

sepalani commented on Source/Core/Common/Assembler/GekkoIRGen.cpp:
```diff
@@ -0,0 +1,829 @@
+void GekkoIRPlugin::PadAlign(u32 bits)
+{
+  const u32 align_mask = (1 << bits) - 1;
+  u32 current_addr = m_active_block->BlockEndAddress();
```
```suggestion
  const u32 current_addr = m_active_block->BlockEndAddress();
```

sepalani reviewed (changes requested):
Parts of this PR are way beyond my area of expertise. So here is a very basic code review. I haven't tested this PR yet but will very soon. Regardless, very impressive work.

vyuuui commented:
That should cover everything.

vyuuui commented on Source/Core/DolphinQt/Debugger/AssemblyEditor.cpp:
```diff
@@ -0,0 +1,330 @@
+    if (base_address.back() == '\n')
+    {
+      base_address.pop_back();
+    }
```
this was unsafe-ish regardless (though i think back just spits back 0 if empty?)

vyuuui commented on Source/Core/Common/Assembler/AssemblerShared.cpp:
```diff
@@ -0,0 +1,27 @@
+  {
+    space_char = '\0';
+  }
+
```
I could have sworn that fmt would just ignore it, as it had been working like this. Regardless, a string should be clearer anyway.

vyuuui commented on Source/Core/DolphinQt/DolphinQt.vcxproj:
```diff
@@ -481,4 +489,4 @@
   </Target>
-</Project>
\ No newline at end of file
+</Project>
```
thanks, vim...

TryTwo commented:
@dolphin-emu-bot rebuild

TryTwo commented:
Unless it was fixed in a recent update. When saving, it puts the entire path into the tab name (rather than just the new saved file's name), making it really long. When loading, it correctly puts just the file name.

vyuuui commented:
> Unless it was fixed in a recent update. When saving, it puts the entire path into the tab name (rather than just the new saved file's name), making it really long. When loading, it correctly puts just the file name.

Fixed, and reminded me that there is no disambiguation for tabs with matching filenames, so I've changed it to display the full path if multiple files have the same name.

vyuuui commented:
Looking more at the current disassembler, it's wildly inaccurate for basically all of the SIMD instructions. It (unnecessarily) includes operations (e.g. `ps_msum pD, pA*pC-pB`), it doesn't show the record bit for SIMD, prefixes registers with `p`, and probably more I'm not immediately seeing.

The choices here are either a ~2k line addition for a brand new disassembler, or patching the existing disassembler to match syntax that this PR expects. Drawbacks to both, and not sure if that's worth doing just to fill the 'Assemble Instruction' dialog box properly, though I'm open to either option if deemed necessary.

TryTwo commented:
I wouldn't mind if PS stuff is left for another PR, or just skipped.

Is there a way to convert a gecko code back into PPC? Like codewrite does.
Also how much is AR supported, is there syntax for conditionals (08)?

vyuuui commented:
Codewrite uses `vdappc`, which is based on the same disassembler that dolphin uses. It looks like it branched from the original implementation which took the route of more human-friendly output. It all comes from the same source, it just looks like someone else did the work there to make it have assembler-friendly syntax.

vyuuui commented:
> Also how much is AR supported, is there syntax for conditionals (08)?

AR only supports direct RAM writes. I'm very open to modifications to the Gecko/AR output, as right now it's very barebones and may behave counterintuitively to what someone inexperienced expects (e.g. referencing absolute addresses with Gecko Exec/Trampoline will *not* do what you want thanks to relocation). I don't think there's a good way to deal with that aside from conjuring up some over-complexified dynamic relocation fixup being emitted alongside the code. My initial idea with the different outputs is more of a convenience factor to format it into a gecko code, and expect the user to be aware they should be writing their asm position independent if needed.

TryTwo commented:
Yeah I think being a stand-in for CodeWrite is a good starting place, then later PRs could do more if there's interest.

I don't think I got/understood your thoughts on having two-way conversions for PowerPC Input <-> Gecko Output in the assembly widget.  Would you have to include vdappc for it to work?

vyuuui commented:
The point of this was to avoid the inclusion of binary blobs such as vdappc or gnu-as. Moreover, vdappc is built to disassemble binaries, not individual instructions. What would have to be done is either copy the work done in vdappc as a separate disassembler (e.g. into externals, but build it as a library), patch the current disassembler to behave like vdappc, or write a new one. The first option is obviously the most preferable, as it has the least added work.

vyuuui commented:
> Yeah I think being a stand-in for CodeWrite is a good starting place, then later PRs could do more if there's interest.
> 
> I don't think I got/understood your thoughts on having two-way conversions for PowerPC Input <-> Gecko Output in the assembly widget. Would you have to include vdappc for it to work?

Oh whoops, I'd misread what you meant, converting Gecko codes back into PPC? I think it just trims off the first two words as well as the trailing `00000000` if present, writes that to a file, then runs *that* through vdappc. Sorry for the misunderstanding.

Tilka commented:
I get icon loading errors when using a non-default theme:
```
  Condition: icon.availableSizes().size() > 0
  File: Source/Core/DolphinQt/Resources.cpp
  Line: 53
  Function: LoadNamedIcon
```

Dentomologist reviewed (commented):
We should probably have an assembler_ prefix on the icon names to avoid namespace clashing or confusion.

vyuuui commented:
> I get icon loading errors when using a non-default theme:
> 

Missed other themes initially, put some placeholders in there that may not fit perfectly with other base themes.

> Other issues (these could be addressed in a future pull request if you prefer):
> 
>     * Unsaved changes are silently lost when closing Dolphin rather than the file tab.

Placed this check the best place I could find, can't attest if that's where it should go er... architecturally?

>     * The editor currently shows dark text on dark background for me (GNOME with dark theme).

This should be dealt with, requires a nasty hack to detect dark mode for custom syntax colors (base color should match system now).

>     * The text size in the editor is rather large. I don't mind that being the default but it's currently not possible to adjust using zoom shortcuts like Ctrl+Mousewheel or Ctrl + +/-/0.

Since the gamelist has a shortcut for that already, its context had to be changed (shouldn't cause issues).

>     * The buttons are probably a bit big. Compare with the buttons in the Breakpoints widget which seem more appropriately sized/padded.

Reduced the size, mostly had them large to get a clear visual on the design.

>     * Maybe the base address field should have some default value instead of being empty?

Just having it fill with 0 for now then.


Other minor changes:
 - Switch output type dropdown to be shorter to fit other themes.
 - Change a few other colors around to account for darker themes.
 - Change unsaved changes dialog to prompt a save, ignore, or cancel

Tilka commented:
- Clicking the Inject button should update the debugger state to avoid showing stale data in the other widgets.
- It would be useful to be able to reference symbols from PPCSymbolDB rather than only labels in the current file.

Tilka commented on Source/Core/DolphinQt/Debugger/AssembleInstructionDialog.cpp:
```diff
@@ -0,0 +1,129 @@
+      m_code = (m_code << 8) | block_bytes[i];
+    }
+
+    m_error_loc_label->setText(tr("<span style=\"color: green; font-weight: bold\">Ok</span>"));
```
"Ok" -> "OK"

Tilka commented:
I've changed a tr() calls to separate HTML formatting from translatable strings. Feel free to squash into your commits if you want.

vyuuui commented:
Is this something I should be worried about?

AdmiralCurtiss commented:
The OSX builder was updated and seems to be producing inconsistent renders, you can ignore it.
